File: zaos/src-tauri/src/events/zaos_events.rs
//! ZAOS Normalized Event Model — provider-neutral.
//!
//! Every runtime (Claude, Codex, …) maps its native events to this model
//! before emitting to the frontend. The frontend never sees runtime-specific types.

use serde::Serialize;
use serde_json::Value;

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type")]
pub enum ZaosEvent {
    // -- Session lifecycle --------------------------------------------------
    #[serde(rename = "session_init")]
    SessionInit {
        #[serde(rename = "sessionId")]
        session_id: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        model: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        tools: Option<Vec<String>>,
        #[serde(rename = "mcpServers", skip_serializing_if = "Option::is_none")]
        mcp_servers: Option<Vec<Value>>,
        #[serde(skip_serializing_if = "Option::is_none")]
        cwd: Option<String>,
        #[serde(rename = "runtimeVersion", skip_serializing_if = "Option::is_none")]
        runtime_version: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        agents: Option<Vec<String>>,
    },

    // -- Streaming text & thinking ------------------------------------------
    #[serde(rename = "message_delta")]
    MessageDelta { text: String },

    #[serde(rename = "thinking_delta")]
    ThinkingDelta { text: String },

    #[serde(rename = "stream_control")]
    StreamControl {
        action: String,
        #[serde(rename = "blockType", skip_serializing_if = "Option::is_none")]
        block_type: Option<String>,
    },

    // -- Complete messages (snapshots) --------------------------------------
    #[serde(rename = "message_complete")]
    MessageComplete {
        #[serde(rename = "messageId")]
        message_id: String,
        text: String,
    },

    #[serde(rename = "thinking_complete")]
    ThinkingComplete {
        #[serde(rename = "messageId")]
        message_id: String,
        text: String,
    },

    // -- Tool calls ---------------------------------------------------------
    #[serde(rename = "tool_call_started")]
    ToolCallStarted {
        id: String,
        name: String,
        input: Value,
        #[serde(rename = "parentToolUseId", skip_serializing_if = "Option::is_none")]
        parent_tool_use_id: Option<String>,
        #[serde(rename = "messageId")]
        message_id: String,
    },

    #[serde(rename = "tool_call_finished")]
    ToolCallFinished {
        #[serde(rename = "toolUseId")]
        tool_use_id: String,
        output: String,
        #[serde(rename = "isError")]
        is_error: bool,
    },

    // -- Permissions --------------------------------------------------------
    #[serde(rename = "approval_requested")]
    ApprovalRequested {
        #[serde(rename = "requestId")]
        request_id: String,
        #[serde(rename = "toolName", skip_serializing_if = "Option::is_none")]
        tool_name: Option<String>,
        #[serde(rename = "toolInput", skip_serializing_if = "Option::is_none")]
        tool_input: Option<Value>,
        #[serde(skip_serializing_if = "Option::is_none")]
        description: Option<String>,
    },

    // -- Run lifecycle ------------------------------------------------------
    #[serde(rename = "run_completed")]
    RunCompleted {
        #[serde(rename = "isError")]
        is_error: bool,
        #[serde(rename = "durationMs")]
        duration_ms: u64,
        #[serde(rename = "durationApiMs")]
        duration_api_ms: u64,
        #[serde(rename = "numTurns")]
        num_turns: u32,
        result: String,
        #[serde(rename = "stopReason")]
        stop_reason: String,
        #[serde(rename = "sessionId")]
        session_id: String,
        #[serde(rename = "costUsd")]
        cost_usd: f64,
    },

    #[serde(rename = "token_usage")]
    TokenUsage {
        #[serde(rename = "inputTokens")]
        input_tokens: u64,
        #[serde(rename = "outputTokens")]
        output_tokens: u64,
        #[serde(rename = "cacheReadTokens")]
        cache_read_tokens: u64,
        #[serde(rename = "durationMs")]
        duration_ms: u64,
        #[serde(rename = "costUsd")]
        cost_usd: f64,
    },

    #[serde(rename = "rate_limited")]
    RateLimited {
        status: String,
        #[serde(rename = "resetsAt")]
        resets_at: u64,
    },

    // -- Agent delegation ---------------------------------------------------
    #[serde(rename = "agent_activity")]
    AgentActivity {
        #[serde(rename = "delegationId")]
        delegation_id: String,
    },

    // -- Catch-all ----------------------------------------------------------
    #[serde(rename = "unknown")]
    Unknown {},
}

File: zaos/src/types/zaosEvents.ts
/**
 * ZAOS Normalized Event Model — provider-neutral.
 *
 * Every runtime (Claude, Codex, ...) must map its native events to this model.
 * Hooks and stores consume only these types, never runtime-specific ones.
 */

// ---------------------------------------------------------------------------
// Discriminated union
// ---------------------------------------------------------------------------

export type ZaosEvent =
  | SessionInitEvent
  | MessageDeltaEvent
  | ThinkingDeltaEvent
  | StreamControlEvent
  | MessageCompleteEvent
  | ThinkingCompleteEvent
  | ToolCallStartedEvent
  | ToolCallFinishedEvent
  | ApprovalRequestedEvent
  | PolicyDecisionEvent
  | RunCompletedEvent
  | TokenUsageEvent
  | RateLimitedEvent
  | AgentActivityEvent
  | UnknownEvent;

// ---------------------------------------------------------------------------
// Session lifecycle
// ---------------------------------------------------------------------------

export interface SessionInitEvent {
  type: "session_init";
  sessionId: string;
  model?: string;
  tools?: string[];
  mcpServers?: unknown[];
  cwd?: string;
  runtimeVersion?: string;
  agents?: string[];
}

// ---------------------------------------------------------------------------
// Streaming text & thinking
// ---------------------------------------------------------------------------

export interface MessageDeltaEvent {
  type: "message_delta";
  text: string;
}

export interface ThinkingDeltaEvent {
  type: "thinking_delta";
  text: string;
}

/** Stream-level flow control signals */
export interface StreamControlEvent {
  type: "stream_control";
  action: "block_start" | "block_stop" | "message_end";
  blockType?: "text" | "thinking" | "tool_use";
}

// ---------------------------------------------------------------------------
// Complete messages (snapshots)
// ---------------------------------------------------------------------------

export interface MessageCompleteEvent {
  type: "message_complete";
  messageId: string;
  text: string;
}

export interface ThinkingCompleteEvent {
  type: "thinking_complete";
  messageId: string;
  text: string;
}

// ---------------------------------------------------------------------------
// Tool calls
// ---------------------------------------------------------------------------

export interface ToolCallStartedEvent {
  type: "tool_call_started";
  id: string;
  name: string;
  input: Record<string, unknown>;
  parentToolUseId?: string;
  messageId: string;
}

export interface ToolCallFinishedEvent {
  type: "tool_call_finished";
  toolUseId: string;
  output: string;
  isError: boolean;
}

// ---------------------------------------------------------------------------
// Permissions
// ---------------------------------------------------------------------------

export interface ApprovalRequestedEvent {
  type: "approval_requested";
  requestId: string;
  toolName?: string;
  toolInput?: Record<string, unknown>;
  description?: string;
  /** Policy Engine fields — set by ZAOS, not by the runtime provider */
  policyVerdict?: "allow" | "ask" | "deny";
  policyRiskLevel?: "low" | "medium" | "high" | "critical";
  policyReason?: string;
  policyMatchedRules?: string[];
}

// ---------------------------------------------------------------------------
// Policy Engine decisions (auto-allow / auto-deny log)
// ---------------------------------------------------------------------------

/** Logged when ZAOS Policy Engine auto-allows or auto-denies an action (not shown as permission prompt) */
export interface PolicyDecisionEvent {
  type: "policy_decision";
  toolName?: string;
  verdict: "allow" | "deny";
  riskLevel: "low" | "medium" | "high" | "critical";
  reason: string;
  matchedRules: string[];
}

// ---------------------------------------------------------------------------
// Run lifecycle
// ---------------------------------------------------------------------------

export interface RunCompletedEvent {
  type: "run_completed";
  isError: boolean;
  durationMs: number;
  durationApiMs: number;
  numTurns: number;
  result: string;
  stopReason: string;
  sessionId: string;
  costUsd: number;
}

export interface TokenUsageEvent {
  type: "token_usage";
  inputTokens: number;
  outputTokens: number;
  cacheReadTokens: number;
  durationMs: number;
  costUsd: number;
}

export interface RateLimitedEvent {
  type: "rate_limited";
  status: string;
  resetsAt: number;
}

// ---------------------------------------------------------------------------
// Agent delegation
// ---------------------------------------------------------------------------

/** Heartbeat: an event was received from a running sub-agent */
export interface AgentActivityEvent {
  type: "agent_activity";
  delegationId: string;
}

// ---------------------------------------------------------------------------
// Catch-all
// ---------------------------------------------------------------------------

export interface UnknownEvent {
  type: "unknown";
}

// ---------------------------------------------------------------------------
// Future event placeholders (not yet emitted by any runtime)
// ---------------------------------------------------------------------------
// These types are defined for forward compatibility.
// They will be added to the ZaosEvent union when implemented.

/** An artifact (file, image, etc.) was created or modified */
export interface ArtifactUpdatedEvent {
  type: "artifact_updated";
  artifactType: "file" | "image" | "diff";
  path?: string;
  description?: string;
}

/** A proof (screenshot, test result, etc.) was added */
export interface ProofAddedEvent {
  type: "proof_added";
  proofType: "screenshot" | "test_result" | "build_output";
  path?: string;
  summary?: string;
}

/** Session summary is ready (end of run) */
export interface SessionSummaryReadyEvent {
  type: "session_summary_ready";
  summary: string;
  tokensTotal: number;
  costTotal: number;
}
